model/metadata/system_test.go

```diff
@@ -18,165 +18,181 @@
 		},
-	}
-
-	for _, test := range tests {
-		output := test.System.fields()
-		assert.Equal(t, test.Output, output)
-	}
-}
-
-func TestSystemDecode(t *testing.T) {
-	host, name, arch, platform, ip := "host", "custom hostname", "amd", "osx", "127.0.0.1"
-	inpErr := errors.New("some error")
-	for _, test := range []struct {
-		input         interface{}
-		inputErr, err error
-		s             *System
-	}{
-		{input: nil, err: nil, s: nil},
-		{input: nil, inputErr: inpErr, err: inpErr, s: nil},
-		{input: "", err: errors.New("invalid type for system"), s: nil},
-		{
-			input: map[string]interface{}{"hostname": 1},
-			err:   utility.ErrFetch,
-			s:     &System{Hostname: nil, Architecture: nil, Platform: nil, IP: nil},
+		"k8s namespace with configured hostname": {
+			input: map[string]interface{}{
+				"kubernetes":          map[string]interface{}{"namespace": namespace},
+				"detected_hostname":   detected,
+				"configured_hostname": configured,
+			},
+			s: &System{Kubernetes: &Kubernetes{Namespace: &namespace}, DetectedHostname: &detected, ConfiguredHostname: &configured},
 		},
-		{
+		"k8s empty": {
 			input: map[string]interface{}{
-				"hostname": host, "architecture": arch, "platform": platform, "ip": ip,
+				"kubernetes":          map[string]interface{}{},
+				"detected_hostname":   detected,
+				"configured_hostname": configured,
 			},
-			err: nil,
-			s:   &System{Hostname: &host, Architecture: &arch, Platform: &platform, IP: &ip},
+			s: &System{Kubernetes: &Kubernetes{}, DetectedHostname: &detected, ConfiguredHostname: &configured},
 		},
-		{
+		"full": {
 			input: map[string]interface{}{
-				"hostname": host, "name": name,
+				"platform":     platform,
+				"architecture": arch,
+				"ip":           ip,
+				"container":    map[string]interface{}{"id": containerID},
+				"kubernetes": map[string]interface{}{
+					"namespace": namespace,
+					"node":      map[string]interface{}{"name": nodename},
+					"pod": map[string]interface{}{
+						"uid":  podUID,
+						"name": podname,
+					},
+				},
+				"hostname":            host,
+				"configured_hostname": configured,
+				"detected_hostname":   detected,
 			},
 			err: nil,
-			s:   &System{Hostname: &host, Name: &name},
+			s: &System{
+				DetectedHostname:   &detected,
+				ConfiguredHostname: &configured,
+				Architecture:       &arch,
+				Platform:           &platform,
+				IP:                 &ip,
+				Container:          &Container{ID: containerID},
+				Kubernetes:         &Kubernetes{Namespace: &namespace, NodeName: &nodename, PodName: &podname, PodUID: &podUID},
+			},
 		},
 	} {
-		sys, err := DecodeSystem(test.input, test.inputErr)
-		assert.Equal(t, test.s, sys)
-		assert.Equal(t, test.err, err)
+
+		t.Run("Decode"+name, func(t *testing.T) {
+			system, err := DecodeSystem(test.input, test.inputErr)
+			assert.Equal(t, test.s, system)
+			assert.Equal(t, test.err, err)
+		})
+
+		if test.err == nil {
+			t.Run("Transform"+name, func(t *testing.T) {
```

Comment:
can this be done within the `Decode` run to avoid decoding twice?  Not a huge deal if there is a reason to separate - would you just add that reason in a comment as this is surprising, to me at least, as is.